### ISSUE TITLE
[JOSS review] some fixes required in the documentation

### ISSUE THREAD

**Issue by dmentipl** [SWIFTSIM/swiftsimio#64]
JOSS review issue: https://github.com/openjournals/joss-reviews/issues/2430

# Summary

First of all, swiftsimio is a really well written Python package, filling a need to analyse and visualise large SWIFT datasets. The particle masking feature is particularly cool. It has excellent internal documentation via docstrings. I like that you use black for standardised formatting.

There are a few relatively minor issues that need addressing detailed below.

# Code

One minor thing is that there is a slightly confusing message: "You do not have numba installed. Please consider installing if you are going to be doing visualisation or indexing large arrays (pip install numba)" when importing. However, I do have numba (v0.48.0) installed. I think `numba.core`, which is the missing import that generates this message, was added in v0.49.1.

I have a small suggestion that it would be useful to add an `environment.yml` file to set up a conda environment for users or for development. E.g. I used this while testing the documentation code examples:

```yaml
name: swiftsimio

channels:
  - conda-forge

dependencies:
  - python>=3.6
  - black
  - h5py
  - ipython
  - matplotlib
  - numba>=0.50
  - numpy
  - pip
  - pykdtree
  - pytest
  - scipy
  - tqdm
  - unyt
  - pip:
    - h5glance
    - py-sphviewer
    - swiftsimio
    - velociraptor
    - wily
```

# Manuscript

The only comments I have about the manuscript are:

- there is a typo in the "Background" section in paragraph 1: "gallaxy";
- a small style-point: I think Python should be capitalised.

# Documentation

There are several (mostly minor) issues with the documentation on readthedocs. (I was working off the stable version v4.0.0; commit hash `16e79ce8`, using the provided dataset from [http://virgodb.cosma.dur.ac.uk/swift-webstorage/IOExamples/cosmological_volume.hdf5](http://virgodb.cosma.dur.ac.uk/swift-webstorage/IOExamples/cosmological_volume.hdf5).)

There are two changes that I suggest are important additions to the documentation.

1. I would like to see the output figures generated by the code examples in the "Visualisation" sections to check if the code is working when I run it.
2. It would be good to include a link to an example "catalogue file" required in the code examples in the "Visualisation/Projection" page.

In what follows I've detailed issues I found while running the example code in the documentation.

### Using metadata

- Where it states "This will output [4.38929954e+26 4.38929954e+26 4.38929954e+26] cm" instead I get: `[142.24751067 142.24751067 142.24751067] Mpc`.
- There is a typo in "(a :mod:datetime object describing when the snapshot was written to disk)". I think you just need to replace ":mod:" with ":py:mod:".

### Reading particle data

- Where it states "will output 3.08567758e+24*cm." instead I get: `Mpc`.
- Where it states "which will output '\\frac{M_\\odot}{\\rm{Mpc}^{3}}'." instead I get: `2.0 \times 10^{30}\ \frac{\rm{kg}}{\rm{Mpc}^{3}}`.

### Named columns

- I think this is not formatted as intended. There seems to be a syntax error. I couldn't really run the code.

### Non-unyt properties

- Where it states "which will output Co-moving mass densities of the particles." instead I get: `AttributeError: 'cosmo_array' object has no attribute 'description'`.

### Spatial-only masking

- In the example, I think `data = load(filename, mask=mask)` should be `sw.data = load(filename, mask=mask)`.

### Full mask

- In the example on the line containing `mask.constrain_mask("gas", "density", 0.4 * density_units, 0.8 * density_units)` I get `AttributeError: 'SWIFTUnits' object has no attribute 'gas'`.

### Writing a subset of snapshot

- In the example on the line containing `sw.subset_writer.write_subset("test_subset.hdf5", mask)` I get `AttributeError: module 'swiftsimio' has no attribute 'subset_writer'`.

## Visualisation — Projection

It would be good to include the images generated by the examples, preferably using the available dataset.

### Rotation

- In the example, a catalogue file is required. Can this be provided?

### Lower-level API

- It would be nice if the example was executable.

### Example

- In the second example, should "project_gas" should be replaced by "slice_gas"?

### Example

- In the first code block, should "slice_gas" be replaced by "render_gas"?
- It seems like there may be a memory leak, or at least my laptop doesn't have enough memory to run the second code block with the dataset provided. The code crashes IPython.

## Command-line Utilities

- Running `swiftsnap` on the provided dataset gives me: `AttributeError: 'str' object has no attribute 'decode'`

**Comment by JBorrow** [comment]
Thank you for such a detailed review. I will respond to your comments with version 4.0.1. Apologies that the sample data was so out of date - we never really use that so it just got forgotten about it.

**Comment by JBorrow** [comment]
I believe that #65 should have addressed all of these comments.

Going through one-by-one:

1. I have enabled the code to work with slightly older versions of `numba`
2. Thanks for the `conda` file, we don't use it but I have added this to the repo so that users of that package management software can develop on `swiftsimio` easier.
3. Manuscript changes have been performed in the `joss-paper` branch.
4. I have updated the example data to be able to work with all of the examples; apologies again that this was so out of date.
5. I have added a velociraptor catalogue to the loading data page for use with the snapshot.
6. I have fixed this issue (we used to use cgs for base units everywhere; that's now fixed).
7. I have fixed the reference to `datetime`
8. Same as above; fixed.
9. Same as above; fixed.
10. Named columns should work with the new example dataset; this older data was not compatible with the new named columns scheme.
11. This was changed when we convinced the `unyt` folks to have `name` attribute in their array - we changed from `description` to `name`. This is updated within the documentation now.
12. This typo has been corrected.
13. Thanks - you discovered a bug! We have fixed this now.
14. We were missing an import in `__init__.py`; this should now be fixed.
15. I have added example images in some cases within the visualisation, and provided the catalogue file for download.
16. I can't think of a nice way of making that example executable without having a random field of particles visualised... For now, I would like to keep it as it is.
17. Yes, `project_gas` has been changed to `slice_gas`
18. Yes, I have made this change.
19. Unfortunately no memory leak, but this is very memory-intensive. I have changed the example to use a 256^3 grid instead of 1024^3 (which is 8 Gb and for the parallel version of this we create N_threads copies, oops!)
20. The CLI `swiftsnap` has been fixed and I've added some more safety checks to stop this happening with other quantities.

I would like to thank you again for your kind words and your fast and detailed review!

**Comment by dmentipl** [comment]
JOSS review issue: https://github.com/openjournals/joss-reviews/issues/2430

> I would like to thank you again for your kind words and your fast and detailed review!

No worries. Happy to help!

Thanks for the changes you've made. That's covered most of the issues.

Unfortunately I can't load the new dataset (at [http://virgodb.cosma.dur.ac.uk/swift-webstorage/IOExamples/cosmo_volume_example.hdf5](http://virgodb.cosma.dur.ac.uk/swift-webstorage/IOExamples/cosmo_volume_example.hdf5)). Running off the swiftsimio master branch, I get the following traceback:

```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~/conda/envs/swiftsimio/lib/python3.8/site-packages/swiftsimio/reader.py in postprocess_header(self)
    269             try:
--> 270                 self.snapshot_date = datetime.strptime(
    271                     self.header["Snapshot date"].decode("utf-8"), "%H:%M:%S %Y-%m-%d %Z"

~/conda/envs/swiftsimio/lib/python3.8/_strptime.py in _strptime_datetime(cls, data_string, format)
    567     format string."""
--> 568     tt, fraction, gmtoff_fraction = _strptime(data_string, format)
    569     tzname, gmtoff = tt[-2:]

~/conda/envs/swiftsimio/lib/python3.8/_strptime.py in _strptime(data_string, format)
    348     if not found:
--> 349         raise ValueError("time data %r does not match format %r" %
    350                          (data_string, format))

ValueError: time data '13:59:18 2020-07-08 BST' does not match format '%H:%M:%S %Y-%m-%d %Z'

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-2-9e3866592fc4> in <module>
----> 1 data = sw.load('cosmo_volume_example.hdf5')

~/conda/envs/swiftsimio/lib/python3.8/site-packages/swiftsimio/__init__.py in load(filename, mask)
     90     """
     91
---> 92     return SWIFTDataset(filename, mask=mask)
     93
     94

~/conda/envs/swiftsimio/lib/python3.8/site-packages/swiftsimio/reader.py in __init__(self, filename, mask)
   1214
   1215         self.get_units()
-> 1216         self.get_metadata()
   1217         self.create_particle_datasets()
   1218

~/conda/envs/swiftsimio/lib/python3.8/site-packages/swiftsimio/reader.py in get_metadata(self)
   1250         """
   1251
-> 1252         self.metadata = SWIFTMetadata(self.filename, self.units)
   1253
   1254         return

~/conda/envs/swiftsimio/lib/python3.8/site-packages/swiftsimio/reader.py in __init__(self, filename, units)
    136         self.get_named_column_metadata()
    137
--> 138         self.postprocess_header()
    139
    140         self.load_particle_types()

~/conda/envs/swiftsimio/lib/python3.8/site-packages/swiftsimio/reader.py in postprocess_header(self)
    276                 # a British (en_GB) machine, and then tried to read on a Dutch
    277                 # machine (nl_NL), this would _not_ work because %c is different.
--> 278                 self.snapshot_date = datetime.strptime(
    279                     self.header["Snapshot date"].decode("utf-8"), "%c\n"
    280                 )

~/conda/envs/swiftsimio/lib/python3.8/_strptime.py in _strptime_datetime(cls, data_string, format)
    566     """Return a class cls instance based on the input string and the
    567     format string."""
--> 568     tt, fraction, gmtoff_fraction = _strptime(data_string, format)
    569     tzname, gmtoff = tt[-2:]
    570     args = tt[:6] + (fraction,)

~/conda/envs/swiftsimio/lib/python3.8/_strptime.py in _strptime(data_string, format)
    347     found = format_regex.match(data_string)
    348     if not found:
--> 349         raise ValueError("time data %r does not match format %r" %
    350                          (data_string, format))
    351     if len(data_string) != found.end():

ValueError: time data '13:59:18 2020-07-08 BST' does not match format '%c\n'
```

A couple of other comments about the docs:

- I suggest replacing "cosmological_volume.hdf5" and "my_snapshot_0000.hdf5" everywhere with "cosmo_volume_example.hdf5" now that you've updated the example dataset. (If that is what is intended.)
- I'm still a bit confused about the "Named columns" section of the "Loading Data" page. The code box does not contain valid Python. Maybe I'm misinterpreting it.

**Comment by JBorrow** [comment]
Thanks, I've hopefully fixed that in version 4.0.2. Sorry about that - we don't have any machines that live in different timezones (with different localisations too) that we can test on easily. I should look into how to do this with travis.

I have replaced the names of all of the files at your request.

I have also changed the Named Columns section of the documentation to be clearer: https://swiftsimio.readthedocs.io/en/latest/loading_data/index.html#named-columns
